### ISSUE TITLE
Godot 4

### ISSUE THREAD

**Issue by mason-stafford** [Godot-Opus/libopus-gdnative-asset#3]
Same plugin for Godot 4 will Rock!

**Comment by goatchurchprime** [comment]
I don't know if it helps, but I was compiling godot_4 under nix and spotted libopus as one of the secondary build dependencies.  

I think it comes in via alsalib -> alsa-plugins -> libjack2 -> libopus.  If it's not cleaned out somewhere in the build process, then it seems it's already shipping with a libopus dependency whether we like it or not.  So we might as well use it.

**Comment by mason-stafford** [comment]
thanks for your helpful suggestion, is it possible for you to provide me your compiled version of Godot 4?
@goatchurchprime

**Comment by mason-stafford** [comment]
@Wavesonics
How about you, don't you have any plan for porting to Godot4?

**Comment by goatchurchprime** [comment]
I've got it working, just for linux.  Also I've downgraded to 4.0 because the microphone works in that version, but not in 4.1 under linux, and it tests okay https://github.com/goatchurchprime/libopus-gdnative/tree/4.1

**Comment by goatchurchprime** [comment]
@meisamsafavi The opus library is designed for streaming.  The encode and decode functions just slice it up into small frames and pack them in. We should look at what could be done to turn this into a robust VOIP streaming system that can handle dropped packets.

There's probably a deep rabbithole to go down to make it good.  For example there's that feature you sometimes get in VOIP systems where when there's been a temporary lag in the data, that is later caught up, the stream plays the buffer at double speed briefly in order to make up the time and not leave this buffering in there permanently.

**Comment by Wavesonics** [comment]
So I've moved on from this project in general, very busy with others.

I'm not currently doing Godot related development.

But I love the idea of bringing this forward to Godot 4, so if you guys are interested in taking the lead on doing that I'd certain accept PRs and publish new releases.

**Comment by Wavesonics** [comment]
> The opus library is designed for streaming.

Absolutely, I had always intended to expose methods for doing true streaming. Currently it's all block based: Take in a finished chunk of audio, and transcode it, and send it.

It really shouldn't be too too much work to get it working in a more stream-centric manner. I even had a Godot 3 proof of concept doing it some where....
